### ISSUE TITLE
Change a Scheduled Timer 

### ISSUE THREAD

**Issue by alexzm1** [gurmitteotia/guflow#21]
Is there any way to change the FireAfter TimeSpan of a Timer after being scheduled?

Here is a code snipped of what we're trying to do 

`

public class TestWorkflow : Workflow {
        //WorkflowDescription setup skiped  
	public TestWorkflow(){

		ScheduleActivity<TriggerTimerActivity>().OnCompletion(_=> Ignore);
		
		ScheduleTimer("TimerTest")
			.When(_=>false)
			.FireAfter(TimeSpan.FromMinutes(2)) //This might be changed by a Workflow Signal
			.OnFired(_ =>
				{
					return FailWorkflow("Failed Cancel Timer", "Failed to cancel TimerTest"); //Executed after timer's TimeSpan (2 minutes in this case)
				})
			.OnCancelled(_ =>
				{
					return CompleteWorkflow("Workflow Execution Completed"); //Not Executed After timer cancellation
				}
			);

	}

	[WorkflowEvent(EventName.Signal)]
	public WorkflowAction OnSignal(WorkflowSignaledEvent @event)
	{
		if("start timer".EqualsIgnoreCase(@event.SignalName))
		{
			//Start the "TimerTest" using the @event.Input value as FireAfter TimeSpan value to setup
		}
		if ("cancel timer".EqualsIgnoreCase(@event.SignalName))
		{
			return CancelRequest.ForTimer(name: "TimerTest"); // Timer is cancelled after returning this
		}
		
		return Ignore;
	}
}
`

**Comment by gurmitteotia** [comment]
Amazon SWF does not lend itself nicely to reschedule the timer. However you can use following approach:
```cs
 public class OrderWorkflow : Workflow
    {
        public OrderWorkflow()
        {
            ScheduleTimer("GracePeriod")
                .FireAfter(GracePeriodTimeout) 
                .OnCancel(RescheduleOrCancelWorkflow); // or you can jump to an activity

            ScheduleActivity<CancelOrder>().AfterTimer("GracePeriod");

            ScheduleActivity<ShipOrder>().AfterTimer("GracePeriod");

            
            ScheduleTimer("GracePeriodReschedule").When(_ => false).OnFired(_ => Jump.ToTimer("GracePeriod"));
        }

        private TimeSpan GracePeriodTimeout(ITimerItem item)
        {
            if (AllSignalEvents.Any(a => a.SignalName == "change_time"))
                return TimeSpan.FromMinutes(20); // new time

            return TimeSpan.FromMinutes(10); //Default timeout
        }

        private WorkflowAction RescheduleOrCancelWorkflow(ITimerItem item)
        {
            if (AllSignalEvents.Any(a => a.SignalName == "change_time"))
            {
                if (Timer("GracePeriod").IsActive)
                    CancelRequest.ForTimer("GracePeriod");
            }
            return Jump.ToActivity<CancelOrder>();
        }
    }
```

It would have been much more readable if SWF has raised DecisionTaskScheduled events on timer cancelled event or it would have allow to cancel and schedule a timer in one response. I will again raise it on SWF forums.

**Comment by gurmitteotia** [comment]
I have think again on it. I think it will be useful to have Timer.Reset() and Timer.Reschedule API available. These API are quite generic and and could be useful in many scenarios. I can see that implementation of these APIs will help in setting up the foundation for [ParallelForEach](https://github.com/gurmitteotia/guflow/issues/5) feature. So in future you will be able to either reschedule or reset the timer as below (For simplicity and to stay focused I have removed other details from example) :

```cs
public class OrderWorkflow : Workflow
    {
        public OrderWorkflow()
        {
            ScheduleTimer("GracePeriod").FireAfter(TimeSpan.FromMinutes(30));
           
            ScheduleActivity<ShipOrder>().AfterTimer("GracePeriod");
        }

       [WorkflowEvent(EventName.Signal)]
        public WorkflowAction Signal(WorkflowSignaledEvent @event)
        {
            var timer = Timer("GracePeriod");
            if (@event.SignalName == "change_time")
                if (timer.IsActive) return timer.Reschedule(TimeSpan.FromMinutes(10));

            return Ignore;
        }
    }
```

**Comment by gurmitteotia** [comment]
I have created a [ticket](https://github.com/gurmitteotia/guflow/issues/22) with initial ideas. Any input will be useful.

**Comment by gurmitteotia** [comment]
This feature is implemented in version [1.2.15](https://www.nuget.org/packages/Guflow/1.2.15).
